### PR TITLE
Large Repositories

### LINKED ISSUE #14: When repositories are to large we should fail gracefully.
Currently if a user attempts to download a repository that is to large for the CC computer it will download all the files it can before throwing an out of space error.

### PR DESCRIPTION
**Pull request by aaronmallen**
### Fixed
* #14 added error handling around repositories that are larger then the amount of freespace available.

resolves #14

### CODE REVIEW

**Comment by aaronmallen** [comment]
@eric-wieser this too is ready for CR :D

File: programs/github [hunk 1]
```diff
@@ -108,8 +108,11 @@ if action == subcommands.clone then
 	local tree = repo:tree(treeName)
 
 	-- download the files
-	print("Downloading:")
 	local totalSize = tree.size
+	if totalSize > fs.getFreeSpace(dest) then
+		return printError("Repository is to large for download!")
```
Comment (eric-wieser) [inline programs/github]: Both of the sizes should be included in the error message here

**Comment by aaronmallen** [comment]
Can confirm they use the same measurements [imgur](https://imgur.com/a/tOaw4)

**Comment by aaronmallen** [comment]
I just realized the test you asked me to run is not the one I ran (derp)... so I still think they are using the same measurements... but the numbers didn't come back exactly the same: [Imgur](http://i.imgur.com/JWL8qyD.png?1)

File: programs/github [hunk 2]
```diff
@@ -108,8 +108,11 @@ if action == subcommands.clone then
 	local tree = repo:tree(treeName)
 
 	-- download the files
-	print("Downloading:")
 	local totalSize = tree.size
+	if totalSize > fs.getFreeSpace(dest) then
+		return printError(("Repository is %s, but only %s are free on this computer. Aborting!"):format(sizeStr(totalSize), sizeStr(fs.getFreeSpace(dest)), with sizeStr(1) == ' byte(s)', sizeStr(2048) == '2.0 KiB'))
```
Comment (aaronmallen) [inline programs/github]: This is wrong... can you re-add you comment @eric-wieser

Comment (eric-wieser) [inline programs/github]: ??? Notice how [my old comment](https://github.com/eric-wieser/computercraft-github/pull/15#discussion_r85009489) had some things not in code blocks!

**Comment by eric-wieser** [comment]
Odd that the numbers aren't the same, but they're close enough that this should stop someone cloning something stupid. Perhaps there should be a warning if the sizes differ by less than 20% as well?

**Comment by aaronmallen** [comment]
A warning after download right?

**Comment by eric-wieser** [comment]
No, because then it's too late - a warning before download, so they know if there's a risk the repo won't fit before waiting

**Comment by aaronmallen** [comment]
That's not possible... I have no way of knowing how much space the repo actually takes on the CC computer until after download...

**Comment by aaronmallen** [comment]
Oh Nvm I miss read your comment.

**Comment by eric-wieser** [comment]
What I mean is something like:

```
local sizeError = 0.2
if githubSize() > freeSpace() then
    error()
else if githubSize() * (1 + sizeError) > freeSpace then
    warn_but_continue()  -- maybe ask for confirmation?
end
```

**Comment by aaronmallen** [comment]
Yeah I'm picking up what your laying down, I just miss read your comment.

File: programs/github (hunk 1 above)
Comment (aaronmallen) [inline programs/github]: I guess I don't understand whats happening here... what does the `with` do I'm not familiar with the syntax here...

**Comment by aaronmallen** [comment]
This one's ready for another go @eric-wieser I'm not sure I did what you wanted with `sizeStr` if not could you provide a full example of what you were going for there.

File: programs/github [hunk 3]
```diff
@@ -79,6 +79,30 @@ if flags.branch and flags.tag then
 	return printError('--branch and --tag cannot both be specified')
 end
 
+
+local function hasEnoughSpace(repoSize, freeSpace)
+	local sizeError = 0.2
+
+	local function warnAndContinue()
+		print("Repository maybe to large to download, attempt anyway?")
```
Comment (eric-wieser) [inline programs/github]: A better message:

    write("Repository may be too large to download, attempt anyway? [Y/n]: ")

File: programs/github [hunk 4]
```diff
@@ -79,6 +79,30 @@ if flags.branch and flags.tag then
...
+		local validAnswers = { y = 'yes', yes = 'yes', n = 'no', no = 'no' }
```
Comment (eric-wieser) [inline programs/github]: Accept empty string as yes

Comment (eric-wieser) [inline programs/github]: Accept empty string as yes

File: programs/github [hunk 5]
```diff
@@ -79,6 +79,30 @@ if flags.branch and flags.tag then
...
+		local input = io.read()
```
Comment (eric-wieser) [inline programs/github]: `:lower()`

File: programs/github [hunk 6]
```diff
@@ -79,6 +79,30 @@ if flags.branch and flags.tag then
...
+		while not validAnswers[input] do
+			print("Please type [y]es or [n]o")
+			input = io.read()
+		end
+		return (validAnswers[input] == 'yes')
```
Comment (eric-wieser) [inline programs/github]: No need for parens here

File: programs/github [hunk 7]
```diff
@@ -108,8 +132,26 @@ if action == subcommands.clone then
 	local tree = repo:tree(treeName)
 
 	-- download the files
-	print("Downloading:")
 	local totalSize = tree.size
+	local freeSpace = fs.getFreeSpace(dest)
+
+	if not hasEnoughSpace(totalSize, freeSpace) then
+		local function sizeStr(bytes)
```
Comment (eric-wieser) [inline programs/github]: This function could be hoisted to the top level, I think

File: programs/github (hunk 6 above)
Comment (aaronmallen) [inline programs/github]: AH will lua recognize I'm trying to return a boolean without the parens?

File: programs/github (hunk 7 above)
Comment (aaronmallen) [inline programs/github]: @eric-wieser CR feedback fixed.

**Comment by aaronmallen** [comment]
A warning after download?

File: programs/github [hunk 8]
```diff
@@ -79,6 +79,41 @@ if flags.branch and flags.tag then
 	return printError('--branch and --tag cannot both be specified')
 end
 
+
+local function hasEnoughSpace(repoSize, freeSpace)
+	local sizeError = 0.2
+
+	local function warnAndContinue()
+		print("Repository may be too large to download, attempt anyway? [Y/n]:")
```
Comment (eric-wieser) [inline programs/github]: I deliberately meant `write` here, to leave the line incomplete

**Comment by aaronmallen** [comment]
tested and ready for trigger pull @eric-wieser

File: programs/github [hunk 9]
```diff
@@ -79,6 +79,43 @@ if flags.branch and flags.tag then
 	return printError('--branch and --tag cannot both be specified')
 end
 
+
+local function hasEnoughSpace(repoSize, freeSpace)
+	-- warn if repository will fill more than 80% of the
+	-- current freeSpace.
+	local sizeError = 0.2
```
Comment (eric-wieser) [inline programs/github]: This comment isn't even true! It's 83%.

I means something like "The value reported by github doesn't match the one reported by CC". Summarize the conclusions in your imgur link in the comment.

**Comment by aaronmallen** [comment]
bump

File: programs/github (hunk 9 above)
Comment (eric-wieser) [inline programs/github]: Because it's actually `1 / (1 + 0.2)`!

**Pull request merged**